File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        // 1. reverse whole string
        reverse(s.begin(), s.end());
        int i=0;
        // main pointrs
        int r=0;
        int l=0;
        int n=s.length();
        while(i<n){
            while(i<n && s[i]!= ' '){
                s[r++]=s[i++];
            }
            if(l<r){
                reverse(s.begin()+l, s.begin()+r);
                s[r]=' ';
                r++;
                l=r;
            }
            i++; //its keep moving non stop
        }
        s=s.substr(0,r-1);
        return s;
    }
};

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    public String reverseWords(String s) {
        // 1. Reverse the whole string
        StringBuilder sb = new StringBuilder(s).reverse();
        int i = 0;
        // Main pointers
        int r = 0;
        int l = 0;
        int n = sb.length();
        char[] chars = sb.toString().toCharArray();

        while (i < n) {
            // Find the end of the current word
            while (i < n && chars[i] != ' ') {
                chars[r++] = chars[i++];
            }
            // If there was a word, reverse it
            if (l < r) {
                reverse(chars, l, r - 1);
                if (r < n) { // Only add a space if within bounds
                    chars[r++] = ' ';
                }
                l = r;
            }
            i++; // Keeps moving non-stop
        }

        // Remove trailing space and return the result
        return new String(chars, 0, r > 0 && chars[r - 1] == ' ' ? r - 1 : r);
    }

    private void reverse(char[] chars, int left, int right) {
        while (left < right) {
            char temp = chars[left];
            chars[left] = chars[right];
            chars[right] = temp;
            left++;
            right--;
        }
    }
}

   